talk/expert/cpp20concepts.tex

```diff
@@ -0,0 +1,199 @@
+      \end{itemize}
+    \end{block}
+    \begin{exampleblock}{With {\it auto}}
+      \scriptsize
+      \begin{cppcode*}{}
+      bool equal( MyFloatingPoint auto e1, MyFloatingPoint auto e2 )
+      { return abs(e1-e2)<std::numeric_limits<T>::epsilon() ; }
+      \end{cppcode*}
+    \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+    \frametitlecpp[20]{Requirements and concepts}
+    \begin{block}{Standard concepts}
+      \begin{itemize}
+        \item Writing a bug-proof concept is an expert task.
+        \item Prefer the ones provided by the standard library.
+        \item Not surprisingly, there is one for floating point numbers.
+      \end{itemize}
+    \end{block}
+    \begin{exampleblock}{With {\it std::}}
+      \scriptsize
+      \begin{cppcode*}{}
+      #include <concepts>
+
+      bool equal( std::floating_point auto e1, std::floating_point auto e2 )
+      { return abs(e1-e2)<std::numeric_limits<T>::epsilon() ; }
+      \end{cppcode*}
+    \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+    \frametitlecpp[20]{Requirements and concepts}
+    \begin{block}{Within a function body}
+      \begin{itemize}
+        \item Concepts are also usable wherever a boolean is expected.
+        \item Since they are evaluated at compile-time, they can appear in a {\it if constexpr} condition.
+      \end{itemize}
+    \end{block}
+    \begin{exampleblock}{With {\it if constexpr}}
+      \scriptsize
+      \begin{cppcode*}{}
+      template< typename T>
+      bool equal( T e1, T e2 )
+      {
+        if constexpr (std::floating_point<T>)
+        { return abs(e1-e2)<std::numeric_limits<T>::epsilon() ; }
+        else
+        { return (e1==e2) ; }
+      }
+      \end{cppcode*}
+    \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+    \frametitlecpp[20]{Requirements and concepts}
+    \begin{block}{Advanced requirements}
+        Rather than the basic requirements above, one may use a \textbf{requires-expression}, able to:
+        \begin{itemize}
+            \item include basic requirements as before,
+            \item list some expressions that must be valid,
+            \item check the return type of some expressions.
+        \end{itemize}
+    \end{block}
+    \begin{exampleblock}{Requires-expression}
+      \scriptsize
+      \begin{cppcode*}{}
+      template< typename T>
+      concept StreamableAndComparableNumber = requires( T v1, T v2 )
+       {
+        requires std::integral<T> || std::floating_point<T> ;
+        std::cout<<v1<<v2 ;
+        { equal(v1,v2) } -> std::convertible_to<bool> ;
+       } ;
+      \end{cppcode*}
+    \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+    \frametitlecpp[20]{Requirements and concepts}
+    \begin{block}{To be remembered}
+      \begin{itemize}
+      \item A template can now {\it requires} properties of its parameters.
+      \item Compiler error messages clearly state which argument does not fulfill which requirement.
```

Comment:
Could we maybe show an example for a compiler error before/after introducing concepts?

Reply:
@sponce @hageboeck I tried to put some (highly repaginated) extracts of compiler error messages. What do you think ?

Reply:
I didn't know what I was asking for. 🙂 Those errors are probably too much to show them in the lecture.

Maybe if we only show the last lines?
```
...
error: no type named 'type' in 'struct std::enable_if<false, void>'
```
vs
```
...
the expression 'is_floating_point_v<T> [with T = int]' evaluated to 'false'
```